sig-and-wg/sig-core/proposals/two-phase-connector-service.md

```diff
@@ -0,0 +1,100 @@
+# Connector service with two phase pairing
+
+Created on 2018-12-17 by Lukasz Szymik (@lszymik).
+
+## Status
+
+Proposed on 2018-12-17.
+
+## Motivation
+
+The Connector service is responsible for establishing a secure connection between a connected application and Kyma runtime. It is achieved by providing the client certificate which is later validated by the Application Connector.
+A client certificate is used for registering application metadata, like APIs, and sending events to Kyma. The registration of application metadata and sending events are separated functionalities. They can be realized using different components deployed on different Kyma clusters.
+Therefore, the connection flow must be extended to support cases where the Application Registry can be separated from the Kyma runtime. 
+
+## Goal
+
+1. The Connector service handles client certificate provisioning for an App Registry connection.
+1. An endpoint for getting metadata information with App Registry URL, Event service URL is exposed and accessible with the use of the client certificate. Metadata can be extended to support automated communication between Kyma and connected applications. Example, the Event service endpoint can be changed, and application configuration must be updated.
+1. The Connector service handles client certificate provisioning for a Kyma runtime using the first client certificate.
+1. The new client certificate can be requested using the previous one (certificate rotation)
+1. The Connector service can trigger the creation of the application custom resource if it doesn't exist yet.
+
+## Suggested solution
+
+Connector service is extended to work with two phases:
+
+  - In the first phase, the Connector service is returning cluster metadata for CSR signing and information about App Registry URL.
+  - In the second phase, the Connector service is returning metadata of the Kyma runtime cluster for which a second client certificate can be obtained.
```

Comment:
I was thinking about it a little bit and I have doubts if the second client certificate will help in any aspect. In the beginning, we wanted to separate registration / metadata calls from actual workload (events). When I started to think about handling security issues (mainly client certificate is compromised), I came to the point that we have always invalidate both certificates. In this case, 2 certificates just bring unnecessary complexity.